### PR TITLE
Integrate `compiler_builtins`

### LINKED ISSUE #69: Easing the setup of `compiler_builtins`
What would it take to make `compiler_builtins` be something that didn't need to manually be installed? There are a few approaches:
  - For rustup to include it as a component, so that installation is easier (and it is automatically updated etc.).
  - That the `rust` repo includes it (moonshot, but for those that need `rust-src` it is much more likely that they need it too).
  - Removing it as a dependency by copying the needed builtins into the tree (the kernel does this for others, too).
  - Removing it as a dependency by not using `alloc` (e.g. if for other reasons we end up with our own `alloc`).

### PR DESCRIPTION
**Pull request by ojeda**
This simplifies onboarding, it is smaller, faster to compile and it is one less dependency to worry about.

It also simplifies the build system a tiny bit since now everything uses the 2018 edition.

Fixes #69.

### CODE REVIEW

**Comment by ojeda** [comment]
Let me know what you think about this. I'm not 100% sure if this may bite us later on (e.g. if we end up having to add many builtins back later on for some reason), but it seems like a good choice at the moment.

@alex was worried in #52 about the complexity for newcomers having to set this one up (apart from `rust-src`), which this takes care of. It also means we don't need to ask `rustup` to add it as a component, etc. (although that would be anyway nice for other users, nevertheless).

**Comment by alex** [comment]
Are these really the only builtins we need?

**Comment by ojeda** [comment]
Rewording of the comment to simplify and fix a typo too.

**Comment by ojeda** [comment]
> Are these really the only builtins we need?

These listed here are the ones we shouldn't need (from what I can tell, at least -- I added the ones using `f32`, `f64`, `i128` and `u128`), but we do need now because some generated code uses them nevertheless (which should not be there to begin with).

The ones that we currently/actually need are none, at least looking at the CI being successful. The ones that we may actually need in the future is not clear -- but we can add them as they come (either here or in the C side).

In other words, in the future, if everything is customized for the kernel, this file would be empty (or with just the very subset actually needed *and* used in our code/compiler *and* not already defined by C).

**Review by alex — APPROVED** [review]
Neat. Good enough for me.

**Comment by ojeda** [comment]
I will leave this open for a couple days -- I'd like to give @nickdesaulniers & others involved with LLVM etc. the chance to tell us whether this looks like a sane enough approach.

**Comment by ojeda** [comment]
@bjorn3 is probably interested in this, too, from the `rustc` side.

File: rust/compiler_builtins.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+define_panicking_intrinsics!("`u128` should not be used", {
+    __ashlti3,
+    __lshrti3,
+    __rust_u128_add,
+    __rust_u128_addo,
+    __rust_u128_mulo,
+    __rust_u128_shlo,
+    __rust_u128_shro,
+    __rust_u128_sub,
+    __rust_u128_subo,
```
Comment (bjorn3) [inline rust/compiler_builtins.rs]: The `__rust_*128_*` intrinsics shouldn't ever be linked against when using the LLVM backend.

Comment (ojeda) [reply rust/compiler_builtins.rs]: Ah, yes, those were the ones you said only cranelift is using, right?

Comment (bjorn3) [reply rust/compiler_builtins.rs]: Correct

Comment (ojeda) [reply rust/compiler_builtins.rs]: Done, thanks.

**Comment by bjorn3** [comment]
This makes sense to me. I think this is unlikely to break in the near future. There isn't much that could change that would break it. `#![compiler_builtins]` could be removed, but I think this is unlikely due to need to handle compiler_builtins specially. The set of required intrinsics could also change, but that is likely to be an LLVM change that would also affect clang.

**Comment by ojeda** [comment]
Thanks @bjorn3 for the analysis -- that is very good to hear and matches what I was expecting.
